var/spack/repos/builtin/packages/papi/package.py

```diff
@@ -30,11 +31,59 @@ class Papi(Package):
     version('5.4.1', '9134a99219c79767a11463a76b0b01a2')
     version('5.3.0', '367961dd0ab426e5ae367c2713924ffb')
 
+    variant('example', default=True, description='Install the example files')
+    variant('cuda', default=False, description='Enable CUDA support')
+    variant('nvml', default=False, description='Enable NVML support')
+    variant('infiniband', default=False, description='Enable Infiniband support')
+    variant('powercap', default=False, description='Enable powercap interface support')
+    variant('rapl', default=False, description='Enable RAPL support')
+    variant('lmsensors', default=False, description='Enable lm_sensors support')
+
+    conflicts('+cuda', when='@:5.6.0')
+    conflicts('+nvml', when='@:5.6.0')
+
+    depends_on('cuda', when='+cuda')
+    depends_on('cuda', when='+nvml')
+    depends_on('lm-sensors', when='+lmsensors')
+
     # Does not build with newer versions of gcc, see
     # https://bitbucket.org/icl/papi/issues/46/cannot-compile-on-arch-linux
     patch('https://bitbucket.org/icl/papi/commits/53de184a162b8a7edff48fed01a15980664e15b1/raw', sha256='64c57b3ad4026255238cc495df6abfacc41de391a0af497c27d0ac819444a1f8', when='@5.4.0:5.6.99%gcc@8:')
 
+    def setup_environment(self, spack_env, run_env):
+        if '^cuda' in self.spec:
+            spack_env.set('CUDA_DIR', self.spec['cuda'].prefix)
+            run_env.prepend_path('LD_LIBRARY_PATH',
+                                 self.spec['cuda'].prefix.extras.CUPTI.lib64)
+
+            for path in self.spec['cuda'].libs.directories + self.rpath:
+                if path.find('/stubs') > -1:
+                    spack_env.remove_path('SPACK_RPATH_DIRS', path)
+                    run_env.remove_path('SPACK_RPATH_DIRS', path)
```

Comment:
You shouldn't have to set `SPACK_RPATH_DIRS`  in `run_env` since this is used to control the Spack compiler wrappers, and `run_env` is for manipulating the variables set in the module file.

Setting `SPACK_RPATH_DIRS` in `spack_env` will definitely have an effect, although I'm curious why it is removed only from the RPATHs and not from `SPACK_LINK_DIRS` as well (in this case the compiler wrapper would include it as a `-L` entry but then no RPATH would be inserted).

Reply:
PAPI depends on CUDA for the `cuda` variant. But only headers are needed, not the libraries, because PAPI loads CUDA as `dlopen("libcuda.so")`. So the RPATHs are essential and `SPACK_LINK_DIRS` irrelevant for PAPI. Note: CUDA has two `libcuda.so` files: one in stubs that errors out and the other in system location that will be visible by `dlopen()` without any RPATHs. The code above simply makes sure that `stubs` is not in RPATHs because if it is, error code will be returned for `cuInit()` and PAPI will drop the component. Incidentally, both NVIDIA and GCC avoid this issue by linking with `libcuda.so.1` when use flag `-lcuda`.

Reply:
> Note: CUDA has two libcuda.so files: one in stubs that errors out and the other in system location that will be visible by dlopen() without any RPATHs.

In this case I think the `cuda` package `libs` function should omit the stubs libraries (there is already similar logic for omitting `compat` libraries - could you extend it to omit `stubs`?): if there is a library of the same name in `/stubs/` then users should only get it when they specifically intend to use it (which is to say maybe it should be an option to retrieve the `stubs` `libcuda.so` although that is not required here).

Reply:
I would be reluctant to changing access to `stubs`. They serve their purpose very well for all packages that depend on CUDA (I use `ginkgo`, `magma`, and `occa` to test it). They allow to build on machines with system CUDA and without system CUDA. Based on a few searches through the repos, only `openmpi` and `papi` uses `dlopen()` to access CUDA. I feel the above RPATH adjustments serve this unique usage.
If `stubs` were "available upon request" then all packages that depend on CUDA would have to be changed. And the new one have to explicitly ask for `stubs` unless they build on top of CUDA supplied by the system in `/usr/lib64`. I think this too much to ask of package writers.